lib/TPP/TileConsumerAndFuseProducers.cpp

```diff
@@ -704,72 +771,27 @@ struct TileConsumerAndFuseProducers
-
-    // Compute the tile sizes for each contraction operations or
-    // use the default one.
-    llvm::DenseMap<Operation *, SmallVector<OpFoldResult>> defaultTiles;
-    for (auto contractionOp : linalgContractionOperations) {
-      auto tiles = getDefaultTileSizes(contractionOp, this->tileSizes);
-      if (failed(tiles)) {
-        LLVM_DEBUG(llvm::dbgs() << "Failed to compute default tile sizes for: "
-                                << contractionOp << "\n");
-        return;
+    int fixedPoint = 3;
+    do {
+      func::FuncOp func = getOperation();
+      IRRewriter rewriter(&getContext());
+      doFusion(rewriter, func, this->tileSizes, this->maxDepth);
+
+      {
+        RewritePatternSet patterns(&ctx);
+        // Fold unit-extent dims for linalg on tensors. Since
+        // `populateFoldUnitExtentDimsViaSlicesPatterns` works only with
+        // linalg.generic we need to generalize first using
+        // `populateLinalgNamedOpsGeneralizationPatterns`.
+        linalg::ControlDropUnitDims options;
+        options.rankReductionStrategy = linalg::ControlDropUnitDims::
+            RankReductionStrategy::ExtractInsertSlice;
+        linalg::populateFoldUnitExtentDimsPatterns(patterns, options);
+        tensor::populateMergeConsecutiveInsertExtractSlicePatterns(patterns);
+        linalg::populateLinalgNamedOpsGeneralizationPatterns(patterns);
+        (void)applyPatternsAndFoldGreedily(getOperation(), std::move(patterns));
       }
-      defaultTiles[contractionOp] =
-          getAsOpFoldResult(rewriter.getI64ArrayAttr(*tiles));
-    }
-
-    for (linalg::LinalgOp contractionOp : linalgContractionOperations) {
-      Operation *consumerOp = contractionOp.getOperation();
-      if (this->startFromLastFusableConsumer)
-        consumerOp = getLastFusableEltWiseConsumer(
-            contractionOp, visitedConsumers, defaultTiles);
-      fusionRoots.insert(consumerOp);
-    }
-    LLVM_DEBUG(llvm::dbgs() << "#fusionRoots: " << fusionRoots.size() << "\n");
-
-    SmallVector<Operation *> allLinalgOps;
-    func->walk(
-        [&](linalg::LinalgOp linalgOp) { allLinalgOps.push_back(linalgOp); });
-
-    // We want to walk operations in reverse order as fusion will likely find
-    // larger cluster of operations to fuse if we go bottom-up.
-    std::reverse(allLinalgOps.begin(), allLinalgOps.end());
-
-    for (Operation *linalgOp : allLinalgOps) {
-      if (fusionRoots.count(linalgOp)) {
-        LLVM_DEBUG(llvm::dbgs() << "\n\n");
-        FailureOr<scf::SCFTileAndFuseResult> fuseAndTileResult =
-            fuseWithEltwise(rewriter, cast<TilingInterface>(linalgOp),
-                            defaultTiles, fusedOps, this->maxDepth);
-        LLVM_DEBUG(llvm::dbgs() << "\n\n");
-        if (succeeded(fuseAndTileResult)) {
-          rewriter.replaceOp(
-              linalgOp,
-              (*fuseAndTileResult).replacements[linalgOp->getResults()[0]]);
-        }
-      }
-    }
+    } while (fixedPoint--);
```

Comment:
Now it runs 4 iterations.
Either `--fixedPoint` or `while {}` loop for clarity.

Reply:
true, my bad.